pkgs/development/python-modules/certifi/2019.11.28.nix

```diff
@@ -0,0 +1,29 @@
+{ lib
+, fetchPypi
+, buildPythonPackage
+}:
+
+buildPythonPackage rec {
+  pname = "certifi";
+  version = "2019.11.28";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "25b64c7da4cd7479594d035c08c2d809eb4aab3a26e5a990ea98cc450c320f1f";
+  };
+
+  pythonImportsCheck = [ "certifi" ];
+
+  # no tests implemented
+  doCheck = false;
+
+  meta = with lib; {
+    homepage = "https://certifi.io/";
+    description = "Python package for providing Mozilla's CA Bundle";
+    license = licenses.isc;
+    maintainers = with maintainers; [ ]; # NixOps team
+    knownVulnerabilities = [
+      "Python 2 supported for certifi ended in 2019 and shipping an outdated certificate store carries a non-negligble risk"
```

Comment:
Can anyone describe an actual threat model based on “shipping an outdated certificate store”? I can't think of any that would actually be a problem for normal users, and non-normal users already have much worse to worry about.

IMO adding `knownVulnerabilities` when there are no actually known vulnerabilities only makes people more used to working around it, especially for a package as advertised as nixops, and so is a net negative for the security of NixOS users.

Now, if there were some specific known vulnerability (eg. “this CA that was there in 2019 was removed after having issued malicious certificates that can still be used”), I would totally understand this item; but IMO as currently described this is lacking motivation and it'd be better to have no knownVulnerabilities than this one.

Reply:
easy. Some CA get's their private key stolen and now the thief can issue new certificates. The answer to this is revocation. unless you have OCSP revocation is done via two ways:
1. certificate revocation lists
2. removing the CA from certificate stores like this one
Hence keeping the certificate store up to date is important. For reference: https://sslmate.com/certspotter/failures

Reply:
Hm, what about re-using the bundle from the `cacert` package instead? This would not only solve this issue but also would consolidate the CA bundle to a single location. For example when writing integration tests involving Python software, one usually needs to not only override `cacert` but also `python*Packages.certifi` to inject testing certificates.

Reply:
The approach suggested by @aszlig may also be applicable to the boto 2.49.0 package, as I believe it has an embedded cacert file that is even further out-of-date: https://raw.githubusercontent.com/boto/boto/2.49.0/boto/cacerts/cacerts.txt. I'm not certain when this embedded cacerts.txt is used by boto versus the one provided by python2Packages.certifi.

If someone can confirm that boto's embedded cacerts.txt is actually used (especially through nixops 1.7), we might want to add a knownVulnerabilities to the boto package as well -- for consistency with python2Packages.certifi. Or, provide updated certificates in both packages. I took a quick look at arch and debian to see if they had updated the certs in their boto 2.49.0 packages, and it doesn't seem that they have. So, maybe I am mistaken.

Reply:
@zeri42 Thank you, your link confirms that as of today there is no known instance of a CA having an issue that would warrant the `knownVulnerabilities` toggle since the version of certifi that is being considered here, and so that reverting updates should not be a security vulnerability :) (though it is only one source, of course)

(And also, even if they were in most cases I'd keep my position: an attacker stealing the private key from a CA is not a reasonable threat model for a normal NixOS user, which is the reason why I wrote my first message)

This being said I'm going to step out of this discussion: I'm only focussed on “`knownVulnerabilities` here would be bad”; there are for sure much better things that could be done by improving stuff in various ways :)